### PR TITLE
Get volume options support for Restclient and glustercli

### PR DESCRIPTION
**Pull request by Madhu-1**
Resolves conflicts and addressed review comments of #572 

Signed-off-by: Madhu Rajanna <mrajanna@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@aravindavk  @prashanthpai  please review

File: glustercli/cmd/volume.go [hunk 1]
```diff
@@ -228,10 +228,34 @@ var volumeDeleteCmd = &cobra.Command{
...
+			}).Error("error getting volume options")
+			failure("Error getting Volume options", err, 1)
+		}
+
+		table := tablewriter.NewWriter(os.Stdout)
+		table.SetHeader([]string{"Name", "Modified", "Value", "DefaultValue", "OptionLevel"})
+		table.SetAlignment(tablewriter.ALIGN_LEFT)
+
+		for _, opt := range opts {
+			modifiedStr := "No"
```
Comment (aravindavk) [inline glustercli/cmd/volume.go]: You can use `formatBoolYesNo(opt.Modified)`

Comment (Madhu-1) [reply glustercli/cmd/volume.go]: done.

**Comment by Madhu-1** [comment]
@aravindavk  @prashanthpai please review

File: glusterd2/commands/volumes/volume-options-get.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func volumeOptionsGetHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+
+	volname := mux.Vars(r)["volname"]
+
+	v, err := volume.GetVolume(volname)
+	if err != nil {
+		if err == errors.ErrVolNotFound {
+			restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
```
Comment (aravindavk) [inline glusterd2/commands/volumes/volume-options-get.go]: HTTP status is confusing here. Option not found vs Volume not found. I think `StatusInternalServerError` itself can be used here. Thoughts?

File: glusterd2/commands/volumes/volume-options-get.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	var resp api.VolumeOptionsGetResp
+
+	for _, xl := range xlator.Xlators() {
+		// TODO Once we have information on supported xlators
+		// per volume type we can filter out these options. For
+		// now return all options
+
+		for _, opt := range xl.Options {
+			var val string
+			modified := false
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: done

File: glusterd2/commands/volumes/volume-options-get.go (hunk 2 above)
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: the volume name is an user input, I fell Not found is the best-suited option if volume not found in ETCD
```
status, err := restutils.ErrToStatusCode(err)
```
This function returns back the status code based on the error.

**Comment by Madhu-1** [comment]
@aravindavk anything pending on this PR?

**Comment by aravindavk** [comment]
> @aravindavk anything pending on this PR?

I will review this PR by today

File: pkg/api/volume_resp.go [hunk 4]
```diff
@@ -73,6 +73,15 @@ type VolumeStatusResp struct {
 	Size   SizeInfo   `json:"size"`
 }
 
+// VolumeOptionGetResp is the response sent for a volume option get request
+type VolumeOptionGetResp struct {
+	OptName      string      `json:"name"`
+	Value        string      `json:"value"`
+	Modified     bool        `json:"modified"`
+	DefaultValue string      `json:"default_value"`
+	OptLevel     OptionLevel `json:"option_level"`
```
Comment (aravindavk) [inline pkg/api/volume_resp.go]: Response can contain string

File: glustercli/cmd/volume.go [hunk 5]
```diff
@@ -230,10 +230,29 @@ var volumeDeleteCmd = &cobra.Command{
...
+			}).Error("error getting volume options")
+			failure("Error getting Volume options", err, 1)
+		}
+
+		table := tablewriter.NewWriter(os.Stdout)
+		table.SetHeader([]string{"Name", "Modified", "Value", "DefaultValue", "OptionLevel"})
+		table.SetAlignment(tablewriter.ALIGN_LEFT)
+
+		for _, opt := range opts {
+			table.Append([]string{opt.OptName, formatBoolYesNo(opt.Modified), opt.Value, opt.DefaultValue, opt.OptLevel.String()})
```
Comment (aravindavk) [inline glustercli/cmd/volume.go]: API should return string here `opt.OptLevel.String()`

File: pkg/api/volume_resp.go (hunk 4 above)
Comment (Madhu-1) [reply pkg/api/volume_resp.go]: done

File: glustercli/cmd/volume.go (hunk 5 above)
Comment (Madhu-1) [reply glustercli/cmd/volume.go]: done

**Comment by aravindavk** [comment]
@prashanthpai please review

**Comment by Madhu-1** [comment]
@atinmu rebased.

File: glusterd2/xlator/utils.go [hunk 6]
```diff
@@ -35,14 +35,14 @@ func (e OptionNotFoundError) Error() string {
 // key should be in the [<graph>].<xlator>.<name> form.
 // Returns an error otherwise.
 func FindOption(k string) (*options.Option, error) {
-	// Interested only in <xlator>.<name> part of the key as optMap is indexed
+	// Interested only in <xlator>.<name> part of the key as OptMap is indexed
```
Comment (prashanthpai) [inline glusterd2/xlator/utils.go]: This function was provided to specifically not export `optMap` outside the package.

File: pkg/api/volume_resp.go [hunk 7]
```diff
@@ -73,6 +73,15 @@ type VolumeStatusResp struct {
 	Size   SizeInfo   `json:"size"`
 }
 
+// VolumeOptionGetResp is the response sent for a volume option get request
+type VolumeOptionGetResp struct {
+	OptName      string `json:"name"`
+	Value        string `json:"value"`
+	Modified     bool   `json:"modified"`
+	DefaultValue string `json:"default_value"`
```
Comment (prashanthpai) [inline pkg/api/volume_resp.go]: Use hyphen, not underscores for json strings.

File: pkg/api/optionlevel_jsonenums.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
+// generated by jsonenums -type=OptionLevel; DO NOT EDIT
```
Comment (prashanthpai) [inline pkg/api/optionlevel_jsonenums.go]: Why is this required ? You have already set the field as string type in response.

File: pkg/api/volopt_level.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	OptionStatusBasic
+	OptionStatusExperimental
+	OptionStatusDeprecated
+	OptionStatusUndefined
+)
+
+func (l OptionLevel) String() string {
+	switch l {
+	case OptionStatusBasic:
+		return "Basic"
```
Comment (prashanthpai) [inline pkg/api/volopt_level.go]: What's returned in API response can be small case.

Comment (Madhu-1) [reply pkg/api/volopt_level.go]: This is an auto-generated file, not a good idea to modify

File: pkg/api/optionlevel_jsonenums.go (hunk 8 above)
Comment (Madhu-1) [reply pkg/api/optionlevel_jsonenums.go]: I have taken up in the middle, I have no idea why this is added, do you want me to remove it?

File: pkg/api/volume_resp.go (hunk 7 above)
Comment (Madhu-1) [reply pkg/api/volume_resp.go]: done

File: glusterd2/xlator/utils.go (hunk 6 above)
Comment (Madhu-1) [reply glusterd2/xlator/utils.go]: done

File: glusterd2/commands/volumes/commands.go [hunk 10]
```diff
@@ -32,6 +32,20 @@ func (c *Command) Routes() route.Routes {
 			HandlerFunc:  volumeExpandHandler},
 		// TODO: Implmement volume reset as
 		// DELETE /volumes/{volname}/options
+		route.Route{
+			Name:         "VolumeOptionGet",
+			Method:       "GET",
+			Pattern:      "/volumes/{volname}/options/{optname}",
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glusterd2/commands/volumes/commands.go]: am not sure why this has added, @aravindavk do we have any use-case?

File: glusterd2/commands/volumes/volume-options-get.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/gorilla/mux"
+)
+
+func volumeOptionsGetHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+
+	volname := mux.Vars(r)["volname"]
+
+	v, err := volume.GetVolume(volname)
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: done

File: glustercli/cmd/volume.go [hunk 12]
```diff
@@ -230,10 +230,29 @@ var volumeDeleteCmd = &cobra.Command{
...
+		opts, err := client.VolumeGet(volname, optname)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"error": err.Error(),
+			}).Error("error getting volume options")
+			failure("Error getting Volume options", err, 1)
+		}
+
+		table := tablewriter.NewWriter(os.Stdout)
+		table.SetHeader([]string{"Name", "Modified", "Value", "DefaultValue", "OptionLevel"})
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glustercli/cmd/volume.go]: done

File: glustercli/cmd/volume.go [hunk 13]
```diff
@@ -230,10 +230,29 @@ var volumeDeleteCmd = &cobra.Command{
...
+
+		volname := args[0]
+		optname := args[1]
+
+		opts, err := client.VolumeGet(volname, optname)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"error": err.Error(),
+			}).Error("error getting volume options")
+			failure("Error getting Volume options", err, 1)
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glustercli/cmd/volume.go]: done

File: glustercli/cmd/volume.go [hunk 14]
```diff
@@ -230,10 +230,29 @@ var volumeDeleteCmd = &cobra.Command{
...
 	Run: func(cmd *cobra.Command, args []string) {
-		volname := cmd.Flags().Args()[0]
-		fmt.Println("GET:", volname)
+
+		volname := args[0]
+		optname := args[1]
+
+		opts, err := client.VolumeGet(volname, optname)
+		if err != nil {
+			log.WithFields(log.Fields{
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply glustercli/cmd/volume.go]: done

File: e2e/volume_ops_test.go [hunk 15]
```diff
@@ -402,25 +402,41 @@ func TestVolumeOptions(t *testing.T) {
...
+		r.NotNil(err)
+
+		err = client.VolumeDelete(volname)
+		r.Nil(err)
+
 		createReq.Options = map[string]string{invalidKey: "on"}
 
 		_, err = client.VolumeCreate(createReq)
 		r.NotNil(err)
+
```
[the remark this answers is not in the document]
Comment (Madhu-1) [reply e2e/volume_ops_test.go]: done

File: glusterd2/commands/volumes/commands.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (aravindavk) [reply glusterd2/commands/volumes/commands.go]: If only one option info is required. This filtering can be done in application side but end up iterating all the xlator.options

File: pkg/api/optionlevel_jsonenums.go (hunk 8 above)
Comment (prashanthpai) [reply pkg/api/optionlevel_jsonenums.go]: Yes, as it's already a string in field of response, you can remove it.

**Comment by Madhu-1** [comment]
@prashanthpai addressed review comments, please review

File: glusterd2/commands/volumes/volume-options-get.go [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	optname, found := mux.Vars(r)["optname"]
+
+	if found {
+		opt, err := xlator.FindOption(optname)
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-options-get.go]: `xlator.FindOption` can return two kind of errors: `InvalidKeyError` and `OptionNotFoundError`. A `404` must be returned only for `OptionNotFoundError` and `400` for `InvalidKeyError`

File: pkg/api/volume_resp.go [hunk 17]
```diff
@@ -73,6 +73,15 @@ type VolumeStatusResp struct {
 	Size   SizeInfo   `json:"size"`
 }
 
+// VolumeOptionGetResp is the response sent for a volume option get request
+type VolumeOptionGetResp struct {
+	OptName      string `json:"name"`
+	Value        string `json:"value"`
+	Modified     bool   `json:"modified"`
+	DefaultValue string `json:"default-value"`
+	OptLevel     string `json:"option-level"`
```
Comment (prashanthpai) [inline pkg/api/volume_resp.go]: `OptionLevel`

File: pkg/api/volopt_level.go [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package api
```
Comment (prashanthpai) [inline pkg/api/volopt_level.go]: This must be defined on original type at `glusterd2/xlator/options/options.go`

File: glusterd2/commands/volumes/volume-options-get.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+func createVolumeOptionGetResp(volinfo *volume.Volinfo, opt *options.Option) *api.VolumeOptionGetResp {
+	var resp api.VolumeOptionGetResp
+
+	for _, xl := range xlator.Xlators() {
+		modified, found := false, false
+		val := opt.DefaultValue
+
+		for _, k := range opt.Key {
+			if val, found = volinfo.Options[xl.ID+"."+k]; found {
+				modified = found
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-options-get.go]: Instead of setting it to value of `found`, although the value is true, it's easier to understand if you explicitly set it to `true`

File: glusterd2/commands/volumes/volume-options-get.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+			var val string
+
+			for _, k := range opt.Key {
+				val = opt.DefaultValue
+				modified := false
+
+				if _, found := volinfo.Options[k]; found {
+					val = volinfo.Options[k]
+					modified = true
+				}
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-options-get.go]: I see an opportunity for code reuse here. `createVolumeOptionsGetResp` can may be call `createVolumeOptionGetResp` in a loop. Let me know what you think.

File: glusterd2/commands/volumes/volume-options-get.go (hunk 16 above)
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: done

File: pkg/api/volopt_level.go (hunk 18 above)
Comment (Madhu-1) [reply pkg/api/volopt_level.go]: removed this file

File: glusterd2/commands/volumes/volume-options-get.go (hunk 19 above)
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: done

File: glusterd2/commands/volumes/volume-options-get.go (hunk 20 above)
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: initially, I just did resolved conflicts of  #572.  now i think there is no code duplication

File: pkg/api/volume_resp.go (hunk 17 above)
Comment (Madhu-1) [reply pkg/api/volume_resp.go]: done

File: glusterd2/commands/volumes/volume-options-get.go [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+		OptName:      optname,
+		Value:        optValue,
+		Modified:     modified,
+		DefaultValue: opt.DefaultValue,
+		OptionLevel:  opt.Level.String(),
+	}
+
+	return &resp
+}
+
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-options-get.go]: Looks cleaner now, good job :+1:

File: glusterd2/commands/volumes/volume-options-get.go [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+	for _, xl := range xlator.Xlators() {
+		// TODO Once we have information on supported xlators
+		// per volume type we can filter out these options. For
+		// now return all options
+		for _, opt := range xl.Options {
+			for _, k := range opt.Key {
+				optName := xl.ID + "." + k
+				volOptRest := createVolumeOptionGetResp(volinfo, opt, optName)
+				resp = append(resp, *volOptRest)
+
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-options-get.go]: nit: lost the extra lines

File: glusterd2/xlator/options/options.go [hunk 23]
```diff
@@ -71,6 +71,21 @@ const (
 	OptionStatusDeprecated
 )
 
+func (l OptionLevel) String() string {
+	switch l {
+	case OptionStatusBasic:
```
Comment (prashanthpai) [inline glusterd2/xlator/options/options.go]: Question: Do you really want the API to return response values starting with caps ? `Basic` or `basic` ? JSON responses are usually all small caps. Let me know what you think.

Comment (Madhu-1) [reply glusterd2/xlator/options/options.go]: I just copy pasted the auto-generated code. 
we have lot of String functions which returns string with Capital letters in gd2 and same as sent as response to restclient.

example from volume response

 ``` "type": "Distributed-Replicate"```
 ```"type": "Brick"```

File: glusterd2/commands/volumes/volume-options-get.go (hunk 22 above)
Comment (Madhu-1) [reply glusterd2/commands/volumes/volume-options-get.go]: done

File: glusterd2/xlator/options/options.go (hunk 23 above)
Comment (prashanthpai) [reply glusterd2/xlator/options/options.go]: Fair enough :+1:

**Comment by prashanthpai** [comment]
retest this please
